# Keras Tutorial

Keras is a popular deep learning framework and it's easy to get started.

In [24]:
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D

BATCH_SIZE = 128
NUM_CLASSES = 10
NUM_EPOCHS = 10

First, we read the mnist data and preprocess them.

In [25]:
# download and load the data (split them between train and test sets)
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# expand the channel dimension
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)
input_shape = (28, 28, 1)

# make the value of pixels from [0, 255] to [0, 1] for further process
x_train = x_train.astype('float32') / 255.
x_test = x_test.astype('float32') / 255.

# convert class vectors to binary class matrics
y_train = keras.utils.to_categorical(y_train, NUM_CLASSES)
y_test = keras.utils.to_categorical(y_test, NUM_CLASSES)

Then, we define the model, object function and optimizer that we use to classify.

In [26]:
# define the model
model = Sequential()
model.add(Conv2D(32,(3,3),activation='relu',input_shape=(28,28,1)))
model.add(MaxPooling2D(2,2))
model.add(Conv2D(64,(3,3),activation='relu'))
model.add(MaxPooling2D(2,2))
model.add(Conv2D(64,(3,3),activation='relu'))
model.add(Flatten())
model.add(Dropout(0.1))
model.add(Dense(64,activation='relu'))
model.add(Dense(NUM_CLASSES,activation='softmax'))

# define the object function, optimizer and metrics
model.compile(optimizer='rmsprop',loss='categorical_crossentropy',metrics=['accuracy'])

Next, we can start to train and evaluate!

In [27]:
import numpy as np
# train
model.fit(x_train,y_train,epochs=NUM_EPOCHS,batch_size=BATCH_SIZE)

# evaluate
#score_train =
#print('Training loss: %.4f, Training accuracy: %.2f%%' % (...))
#score_test = 
Testing_loss,Testing_accuracy=model.evaluate(x_test,y_test)
print('Testing loss: %.4f, Testing accuracy: %.2f%%' % (Testing_loss,100*Testing_accuracy))

Epoch 1/10
469/469 [==============================] - 11s 24ms/step - loss: 0.2477 - accuracy: 0.9230
Epoch 2/10
469/469 [==============================] - 11s 24ms/step - loss: 0.0617 - accuracy: 0.9810
Epoch 3/10
469/469 [==============================] - 11s 24ms/step - loss: 0.0419 - accuracy: 0.9864
Epoch 4/10
469/469 [==============================] - 11s 24ms/step - loss: 0.0320 - accuracy: 0.9902
Epoch 5/10
469/469 [==============================] - 11s 24ms/step - loss: 0.0261 - accuracy: 0.9918
Epoch 6/10
469/469 [==============================] - 11s 24ms/step - loss: 0.0212 - accuracy: 0.9933
Epoch 7/10
469/469 [==============================] - 11s 24ms/step - loss: 0.0181 - accuracy: 0.9943
Epoch 8/10
469/469 [==============================] - 11s 24ms/step - loss: 0.0151 - accuracy: 0.9954
Epoch 9/10
469/469 [==============================] - 12s 25ms/step - loss: 0.0129 - accuracy: 0.9955
Epoch 10/10
313/313 [==============================] - 1s 2ms/step - loss: 0.0311 

#### Q6:
Please print the training and testing accuracy.

|                   | loss   | accuracy |
|-------------------|--------|----------|
| train(last epoch) | 0.0123 | 0.9961   |
| test              | 0.0311 | 0.9918   |